processor/batchprocessor/splitmetrics.go

```diff
@@ -19,56 +19,53 @@ import (
 )
 
 // splitMetrics removes metrics from the input data and returns a new data of the specified size.
-func splitMetrics(size int, toSplit pdata.Metrics) pdata.Metrics {
-	if toSplit.MetricCount() <= size {
-		return toSplit
+func splitMetrics(size int, src pdata.Metrics) pdata.Metrics {
+	if src.MetricCount() <= size {
+		return src
 	}
-	copiedMetrics := 0
-	result := pdata.NewMetrics()
-	result.ResourceMetrics().Resize(toSplit.ResourceMetrics().Len())
-	rms := toSplit.ResourceMetrics()
+	totalCopiedMetrics := 0
+	dest := pdata.NewMetrics()
 
-	rmsCount := 0
-	for i := rms.Len() - 1; i >= 0; i-- {
-		rmsCount++
-		rm := rms.At(i)
-		destRs := result.ResourceMetrics().At(result.ResourceMetrics().Len() - 1 - i)
-		rm.Resource().CopyTo(destRs.Resource())
-
-		destRs.InstrumentationLibraryMetrics().Resize(rm.InstrumentationLibraryMetrics().Len())
+	src.ResourceMetrics().RemoveIf(func(srcRm pdata.ResourceMetrics) bool {
+		// If we are done skip everything else.
+		if totalCopiedMetrics == size {
+			return false
+		}
 
-		ilmCount := 0
-		for j := rm.InstrumentationLibraryMetrics().Len() - 1; j >= 0; j-- {
-			ilmCount++
-			instMetrics := rm.InstrumentationLibraryMetrics().At(j)
-			destInstMetrics := destRs.InstrumentationLibraryMetrics().At(destRs.InstrumentationLibraryMetrics().Len() - 1 - j)
-			instMetrics.InstrumentationLibrary().CopyTo(destInstMetrics.InstrumentationLibrary())
+		destRs := dest.ResourceMetrics().AppendEmpty()
+		srcRm.Resource().CopyTo(destRs.Resource())
 
-			if size-copiedMetrics >= instMetrics.Metrics().Len() {
-				destInstMetrics.Metrics().Resize(instMetrics.Metrics().Len())
-			} else {
-				destInstMetrics.Metrics().Resize(size - copiedMetrics)
-			}
-			for k, destIdx := instMetrics.Metrics().Len()-1, 0; k >= 0 && copiedMetrics < size; k, destIdx = k-1, destIdx+1 {
-				metric := instMetrics.Metrics().At(k)
-				metric.CopyTo(destInstMetrics.Metrics().At(destIdx))
-				copiedMetrics++
-				// remove metric
-				instMetrics.Metrics().Resize(instMetrics.Metrics().Len() - 1)
+		srcRm.InstrumentationLibraryMetrics().RemoveIf(func(srcIm pdata.InstrumentationLibraryMetrics) bool {
+			// If we are done skip everything else.
+			if totalCopiedMetrics == size {
+				return false
```

Comment:
The `return true` and `return false` operations for RemoveIf callbacks here are a bit difficult to understand (`return false` is not self-explanatory). Please add comments to explain what is happening.
A nicer approach perhaps would be to change RemoveIf's callback to return an enum instead of bool. In that case we could write e.g. `return pdata.Remove`, `return pdata.Keep` or `return pdata.BreakIteration` or similar.

Reply:
I like it, the "problem" is that we will avoid calling the callback but we still need to iterate over all to shift them to the left in case something was removed.